test/plugin/base_test.rb

```diff
@@ -0,0 +1,1308 @@
+        pod_name: CONTAINER_POD_NAME,
+        container_name: CONTAINER_CONTAINER_NAME,
+        labels: {
+          CONTAINER_LABEL_KEY => CONTAINER_LABEL_VALUE
+        }
+      }
+    }
+  end
+
+  def container_log_entry(log, stream = CONTAINER_STREAM)
+    {
+      log: log,
+      stream: stream,
+      time: CONTAINER_TIMESTAMP
+    }
+  end
+
+  def cloudfunctions_log_entry(i)
+    {
+      stream: 'stdout',
+      log: '[D][2015-09-25T12:34:56.789Z][123-0] ' + log_entry(i)
+    }
+  end
+
+  def cloudfunctions_log_entry_text_not_matched(i)
+    {
+      stream: 'stdout',
+      log: log_entry(i)
+    }
+  end
+
+  def log_entry(i)
+    'test log entry ' + i.to_s
+  end
+
+  def check_labels(entry, common_labels, expected_labels)
+    # TODO(salty) test/handle overlap between common_labels and entry labels
+    all_labels ||= common_labels
+    all_labels.merge!(entry['metadata']['labels'] || {})
+    all_labels.each do |key, value|
+      assert value.is_a?(String), "Value #{value} for label #{key} " \
+        'is not a string: ' + value.class.name
+      assert expected_labels.key?(key), "Unexpected label #{key} => #{value}"
+      assert_equal expected_labels[key], value, 'Value mismatch - expected ' \
+        "#{expected_labels[key]} in #{key} => #{value}"
+    end
+    assert_equal expected_labels.length, all_labels.length, 'Expected ' \
+      "#{expected_labels.length} labels, got #{all_labels.length}"
+  end
+
+  # The caller can optionally provide a block which is called for each entry.
+  def verify_json_log_entries(n, params, payload_type = 'textPayload')
+    i = 0
+    @logs_sent.each do |batch|
+      batch['entries'].each do |entry|
+        unless payload_type.empty?
+          assert entry.key?(payload_type), 'Entry did not contain expected ' \
+            "#{payload_type} key: " + entry.to_s
+          # Check the payload for textPayload, otherwise it's up to the caller.
+          if payload_type == 'textPayload'
+            assert_equal "test log entry #{i}", entry['textPayload'], batch
+          end
+        end
+
+        assert_equal params[:zone], entry['metadata']['zone']
+        assert_equal params[:service_name], entry['metadata']['serviceName']
+        check_labels entry, batch['commonLabels'], params[:labels]
+        yield(entry) if block_given?
+        i += 1
+        assert i <= n, "Number of entries #{i} exceeds expected number #{n}"
+      end
+    end
+    assert i == n, "Number of entries #{i} does not match expected number #{n}"
+  end
+
+  # This module expects the methods below to be overridden.
+
+  # Create a Fluentd output test driver with the Google Cloud Output plugin.
+  def create_driver(_conf = APPLICATION_DEFAULT_CONFIG, _tag = 'test')
+    _undefined
+  end
+
+  # Set up http or grpc stubs to mock the external calls.
+  def setup_logging_stubs
+    _undefined
+  end
+
+  # Verify the number and the concent of the log entries match the expectation.
```

Comment:
Oh god... I need to learn to spell, or at least read what I wrote one more time...

Changed.